TensorIteratorConfig: Check memory overlap by default

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43423 Check for internal memory overlap in some indexing-type functions
* **#43422 TensorIteratorConfig: Check memory overlap by default**

Differential Revision: [D23298653](https://our.internmc.facebook.com/intern/diff/D23298653)

peterbell10 committed:
TensorIteratorConfig: Check memory overlap by default

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorIteratorConfig: Check memory overlap by default"




[ghstack-poisoned]

peterbell10 committed:
Update on "TensorIteratorConfig: Check memory overlap by default"


Differential Revision: [D23298653](https://our.internmc.facebook.com/intern/diff/D23298653)

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorIteratorConfig: Check memory overlap by default"


Differential Revision: [D23298653](https://our.internmc.facebook.com/intern/diff/D23298653)

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorIteratorConfig: Check memory overlap by default"


Differential Revision: [D23298653](https://our.internmc.facebook.com/intern/diff/D23298653)

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorIteratorConfig: Check memory overlap by default"


Differential Revision: [D23298653](https://our.internmc.facebook.com/intern/diff/D23298653)

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -118,6 +118,7 @@ Tensor & _cat_out_cpu(Tensor& result, TensorList tensors, int64_t dim) {
         "unsupported operation: the input tensors cannot refer to any of the "
         "output memory locations. Found overlap in input tensor ", i);
   }
+  at::assert_no_internal_overlap(result);
```
Add a test for `torch.cat(..., out=blah)`?

peterbell10 replied:
The test was actually in the next PR up the stack. I've moved cat down into this PR because it already had some overlap checking already but otherwise moved most of the manual asserts up to the next PR where the corresponding test is.

zou3519 commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -393,7 +393,6 @@ Tensor& logical_not_(Tensor& self) {
 Tensor& logical_not_out(Tensor& result, const Tensor& self) {
   TensorIterator iter = TensorIteratorConfig()
     .check_all_same_dtype(false)
-    .set_check_mem_overlap(true)
```
nit: signbit_out below also checks memory overlap, so we can delete that line. (line 413/414)

peterbell10 replied:
Had a quick grep for `.set_check_mem_overlap(true)` outside of `TensorIterator`.

zou3519 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -694,6 +699,10 @@ static Tensor & masked_select_out_impl_cpu(Tensor & result, const Tensor & self,
+  at::assert_no_internal_overlap(result);
+  at::assert_no_partial_overlap(result, self);
+  at::assert_no_partial_overlap(result, mask);
+
```
This file also adds memory overlap support for:
- index_select out= variant (we should add a test for this)
- masked_fill_ (we should add a test for this)
- masked_select out=variant (we should add a test for this)

zou3519 replied:
@peterbell10 did you want to add tests for these? I'm fine without them but it is nice to have them to check parity between CPU/CUDA (and to detect regressions)

peterbell10 replied:
They have tests in #43423 and I moved the asserts to that PR as well.

zou3519 replied:
Got it, I didn't realize. Thanks for clarifying

zou3519 reviewed (commented):
Still working my way through a list of all TensorIteratorConfig() callsites. I am around 30% done through that list. Here are some initial comments

peterbell10 committed:
Update on "TensorIteratorConfig: Check memory overlap by default"


Differential Revision: [D23298653](https://our.internmc.facebook.com/intern/diff/D23298653)

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorIteratorConfig: Check memory overlap by default"


Differential Revision: [D23298653](https://our.internmc.facebook.com/intern/diff/D23298653)

[ghstack-poisoned]

zou3519 commented:
https://github.com/pytorch/pytorch/blob/a860be898ee0fbafac14bf68e920187a76ded70d/aten/src/ATen/native/FunctionOfAMatrixUtils.cpp#L86

I think we shouldn't check memory overlap here. It looks like the output is being restrided so that it has a zero stride, so that would trigger the memory overlap check always when this function is called.

zou3519 commented on aten/src/ATen/native/cuda/ScatterGatherKernel.cu:
```diff
@@ -2,6 +2,7 @@
 
 #include <ATen/ATen.h>
 #include <ATen/Dispatch.h>
+#include <ATen/MemoryOverlap.h>
```
We don't need to add this include for `.set_check_mem_overlap(false)` to work

peterbell10 replied:
I've moved this into the correct PR.

zou3519 reviewed (commented):
OK, done reading through all callsites of TensorIteratorConfig. I had some comments above, other than that, this looks good

peterbell10 committed:
Update on "TensorIteratorConfig: Check memory overlap by default"


Differential Revision: [D23298653](https://our.internmc.facebook.com/intern/diff/D23298653)

[ghstack-poisoned]

peterbell10 committed:
Update on "TensorIteratorConfig: Check memory overlap by default"


Differential Revision: [D23298653](https://our.internmc.facebook.com/intern/diff/D23298653)

[ghstack-poisoned]

albanD commented:
BC-breaking note:
Both PRs in this stack are BC-breaking because:
1. they fix silent correctness errors; something that used to be silently incorrect now errors out
2. Any external users of TensorIterator now always get the memory overlap check
